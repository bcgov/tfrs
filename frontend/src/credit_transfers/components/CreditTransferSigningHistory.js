import React, { Component } from 'react'
import PropTypes from 'prop-types'
import moment from 'moment-timezone'

import { CREDIT_TRANSFER_STATUS, CREDIT_TRANSFER_TYPES, LCFS_COMPLIANCE_START_DT } from '../../../src/constants/values'
import { arrayMove, transformTransactionStatusDesc } from '../../utils/functions'

class CreditTransferSigningHistory extends Component {
  static recordedFound (histories) {
    return histories.find(history => (
      history.status.id === CREDIT_TRANSFER_STATUS.recorded.id
    ))
  }

  constructor (props) {
    super(props)

    const recordedFound = CreditTransferSigningHistory.recordedFound(props.history)

    if (recordedFound &&
      moment(props.tradeEffectiveDate).format('YYYYMMDD') < moment(recordedFound.createTimestamp).format('YYYYMMDD')) {
      const approvedIndex = props.history.findIndex(history => (
        history.status.id === CREDIT_TRANSFER_STATUS.approved.id))

      const recordedIndex = props.history.findIndex(history => (
        history.status.id === CREDIT_TRANSFER_STATUS.recorded.id
      ))

      arrayMove(props.history, approvedIndex, recordedIndex)
    }
  }

  _renderApproved (history) {
    // if "recorded" status was found, this means this credit trade
    // was from the historical data entry
    // show "the Director" at all times
    // use effective date as well
    const approveTimeStamp = history.createTimestamp >= moment('2024-01-01')
    return (
      <li key={history.createTimestamp}>
        <strong className="text-success">
        {transformTransactionStatusDesc(history.status.id, history.creditTrade.type.id, history.createTimestamp)} {' '}
        </strong>
        <span>
          on {moment(history.createTimestamp).format('LL')} by the
          <strong> Director </strong> under the
        </span>
        <em>
          {approveTimeStamp ? ' Low Carbon Fuels Act' : ' Greenhouse Gas Reduction (Renewable and Low Carbon Fuel Requirements) Act'}
        </em>
      </li>
    )
  }

  _renderDeclined (history) {
    let roleDisplay = null

    if (history.userRole) {
      roleDisplay = history.userRole.description

      if (history.userRole.name === 'GovDeputyDirector' ||
      history.userRole.name === 'GovDirector') {
        roleDisplay = roleDisplay.replace('Government ', '')
      }
    }

    // if "recorded" status was found, this means this credit trade
    // was from the historical data entry
    // don't show the name and just put in "the {role}" instead

    return (
      <li key={history.createTimestamp}>
        <strong className="text-danger">{transformTransactionStatusDesc(history.status.id, history.creditTrade.type.id, history.createTimestamp)} </strong>
        {CreditTransferSigningHistory.recordedFound(this.props.history) &&
          <span>
            on {moment(this.props.tradeEffectiveDate).format('LL')} by the
            <strong> Director </strong> under the
          </span>
        }
        {!CreditTransferSigningHistory.recordedFound(this.props.history) &&
          <span>
            on {moment(history.createTimestamp).format('LL')} by
            <strong> {history.user.firstName} {history.user.lastName}</strong>
            <strong>{roleDisplay && `, ${roleDisplay}`} </strong> under the
          </span>
        }
        <em> Low Carbon Fuels Act</em>
      </li>
    )
  }

  static _isTransferType (history) {
    if (history.creditTrade && [CREDIT_TRANSFER_TYPES.buy.id, CREDIT_TRANSFER_TYPES.sell.id].indexOf(history.creditTrade.type.id) >= 0) {
      return true
    }
    return false
  }

  static renderAccepted (history) {
    return (<strong>{CreditTransferSigningHistory._isTransferType(history) ? 'Signed and submitted' : 'Signed'}</strong>)
  }

  static renderHistory (history) {
    return (<strong>{history.status.status} </strong>)
  }

  static renderNotRecommended (history) {
    const inputtedDate = new Date(history.createTimestamp)
    return (
      (CreditTransferSigningHistory._isTransferType(history) && inputtedDate >= LCFS_COMPLIANCE_START_DT)
        ? <span><strong>Recommended refusing</strong></span>
        : <span>
        <strong>Reviewed</strong> and
        <strong className="text-danger"> Not Recommended</strong>
      </span>
    )
  }

  static renderRecommended (history) {
    const inputtedDate = new Date(history.createTimestamp)
    return (
      (CreditTransferSigningHistory._isTransferType(history) && inputtedDate >= LCFS_COMPLIANCE_START_DT)
        ? <span><strong>Recommended recording</strong></span>
        : <span>
        <strong>Reviewed</strong> and
        <strong className="text-success"> Recommended</strong>
      </span>
    )
  }

  static renderRecorded (history) {
    history.createTimestamp >= moment('2024-01-01')
    return (history.createTimestamp >= moment('2024-01-01') ? <strong>Created</strong> : <strong>Recorded</strong>)
  }

  static renderRefused (history) {
    return (<span><strong className="text-danger">{CreditTransferSigningHistory._isTransferType(history) ? 'Declined' : 'Refused'}</strong></span>)
  }

  static renderRescinded () {
    return (<strong className="text-danger">Rescinded</strong>)
  }

  static renderSubmitted (history) {
    return (<span><strong>{CreditTransferSigningHistory._isTransferType(history) ? 'Signed and sent' : 'Proposed'}</strong></span>)
  }

  static monthsBetween (date1, date2) {
    let months = (date2.getFullYear() - date1.getFullYear()) * 12
    months -= date1.getMonth()
    months += date2.getMonth()
    if (date2.getDate() < date1.getDate()) {
      months--
    }
    return months
  }

  /**
   * Calculates the credit transfer category and the next change in months.
   *
   * @param {string} agreementDate - The date of the agreement
   * @param {string} submissionDate - The date of submission
   * @param {boolean} categoryDSelected - If true, category 'D' is chosen, ignoring other rules
   * @param {string|null} selectedCategory - A specific category, if provided
   * @returns {Object} The category and time (in months) until the next change, if any!
   */
  static calculateTransferCategoryAndNextChange (agreementDate, submissionDate, categoryDSelected, selectedCategory) {
    // If category 'D' is already chosen, return it immediately
    if (categoryDSelected) {
      return { category: 'D', nextChangeInMonths: null }
    }

    // If a specific category is selected, use it to determine the next change
    if (selectedCategory !== null) {
      const nextChange = (selectedCategory === 'A') ? 6 : (selectedCategory === 'B') ? 12 : null
      return { category: selectedCategory, nextChangeInMonths: nextChange }
    }

    // Otherwise, calculates based on the time since the agreement
    const now = new Date()
    submissionDate = submissionDate < now ? submissionDate : now
    const differenceInMonths = CreditTransferSigningHistory.monthsBetween(new Date(agreementDate), new Date(submissionDate))
    if (differenceInMonths < 6) {
      return { category: 'A', nextChangeInMonths: 6 }
    } else if (differenceInMonths < 12) {
      return { category: 'B', nextChangeInMonths: 12 }
    } else {
      return { category: 'C', nextChangeInMonths: null }
    }
  }

  _renderCategoryHistory () {
    const { history, dateOfWrittenAgreement, tradeCategory, categoryDSelected } = this.props
    // if there is no agreement date, it means this credit transfer
    // was created before we had this field as not optional
    // We won't show categorization if there is no agreement date.
    if (!dateOfWrittenAgreement) {
      return (<></>)
    }

    const lastHistoryItem = history[history.length - 1]
    const createdByGov = history[0].creditTrade?.initiator?.id === 1
    if (history.length > 0 && !createdByGov) {
      const agreementDate = dateOfWrittenAgreement || history[0].createTimestamp
      const selectedCategory = tradeCategory?.category || null
      const { category, nextChangeInMonths } = CreditTransferSigningHistory
        .calculateTransferCategoryAndNextChange(agreementDate, history[0].createTimestamp, categoryDSelected, selectedCategory)
      let nextChangeDate = null
      if (nextChangeInMonths !== null) {
        nextChangeDate = moment(agreementDate).add(nextChangeInMonths, 'months').format('LL')
      }
      const endDate = nextChangeDate
      const categoryCorD = categoryDSelected || category === 'C'
      return (
        <p>
          <li>
            <span>Date of written agreement reached between the two organizations: </span>
            <strong>{moment(agreementDate).format('LL')}</strong>
            {lastHistoryItem.status.id === CREDIT_TRANSFER_STATUS.approved.id &&
              <span> (<strong>Category {category}</strong>)</span>
            }
            {lastHistoryItem.status.id !== CREDIT_TRANSFER_STATUS.approved.id &&
              lastHistoryItem.status.id !== CREDIT_TRANSFER_STATUS.rescinded.id &&
              lastHistoryItem.status.id !== CREDIT_TRANSFER_STATUS.refused.id &&
            (
              <>
                <span> (proposal falls under <strong>Category {category}</strong>{categoryCorD ? ')' : ''}</span>
                {nextChangeDate && (<span> if approved by: <strong>{endDate}</strong>)</span>)}
              </>
            )}
          </li>
        </p>
      )
    }
  }

  render () {
    return (
      <div className="credit-transfer-signing-history">
        <h3 className="signing-authority-header" key="header">Transaction History</h3>
        {this._renderCategoryHistory()}
        {this.props.history.length > 0 &&
        this.props.history.map((history, index, arr) => {
          let action
          if (history.isRescinded) {
            action = CreditTransferSigningHistory.renderRescinded()
          } else {
            if ([ // if the next history entry is also recommended/not recommended, don't show it
              CREDIT_TRANSFER_STATUS.notRecommended.id,
              CREDIT_TRANSFER_STATUS.recommendedForDecision.id,
              CREDIT_TRANSFER_STATUS.recorded.id
            ].includes(history.status.id) &&
            (index + 1) < arr.length && [
              CREDIT_TRANSFER_STATUS.notRecommended.id,
              CREDIT_TRANSFER_STATUS.recommendedForDecision.id,
              CREDIT_TRANSFER_STATUS.recorded.id
            ].includes(arr[index + 1].status.id)) {
              return false
            }

            switch (history.status.id) {
              case CREDIT_TRANSFER_STATUS.accepted.id:
                action = CreditTransferSigningHistory.renderAccepted(history)
                break

              case CREDIT_TRANSFER_STATUS.approved.id:
                return this._renderApproved(history)

              case CREDIT_TRANSFER_STATUS.declinedForApproval.id:
                return this._renderDeclined(history)

              case CREDIT_TRANSFER_STATUS.notRecommended.id:
                action = CreditTransferSigningHistory.renderNotRecommended(history)
                break

              case CREDIT_TRANSFER_STATUS.proposed.id:
                action = CreditTransferSigningHistory.renderSubmitted(history)
                break

              case CREDIT_TRANSFER_STATUS.recommendedForDecision.id:
                action = CreditTransferSigningHistory.renderRecommended(history)
                break

              case CREDIT_TRANSFER_STATUS.recorded.id:
                action = CreditTransferSigningHistory.renderRecorded(history)
                break

              case CREDIT_TRANSFER_STATUS.refused.id:
                action = CreditTransferSigningHistory.renderRefused(history)
                break

              default:
                action = CreditTransferSigningHistory.renderHistory(history)
            }
          }

          return (
            <p key={history.createTimestamp + index}><li>{action} <span> on </span>
              {moment(history.createTimestamp).format('LL')}
              <span> by </span>
              <strong> {history.user.firstName} {history.user.lastName}</strong> of
              <strong> {history.user.organization.name} </strong>
              </li>
            </p>
          )
        })}
      </div>
    )
  }
}

CreditTransferSigningHistory.defaultProps = {
  history: [],
  signatures: [],
  tradeEffectiveDate: null,
  dateOfWrittenAgreement: null,
  tradeCategory: null,
  categoryDSelected: false
}

CreditTransferSigningHistory.propTypes = {
  history: PropTypes.arrayOf(PropTypes.shape({
    createTimestamp: PropTypes.string,
    creditTradeUpdateTime: PropTypes.string,
    isRescinded: PropTypes.bool,
    status: PropTypes.shape({
      id: PropTypes.number,
      status: PropTypes.string
    }),
    user: PropTypes.shape({
      displayName: PropTypes.string,
      firstName: PropTypes.string,
      id: PropTypes.number,
      lastName: PropTypes.string
    }),
    creditTrade: PropTypes.shape({
      initiator: PropTypes.shape({
        id: PropTypes.number
      })
    })
  })),
  signatures: PropTypes.arrayOf(PropTypes.shape({
    createTimestamp: PropTypes.string,
    user: PropTypes.shape({
      displayName: PropTypes.string,
      firstName: PropTypes.string,
      id: PropTypes.number,
      lastName: PropTypes.string,
      organization: PropTypes.shape({
        id: PropTypes.number,
        name: PropTypes.string
      })
    })
  })),
  tradeEffectiveDate: PropTypes.string,
  dateOfWrittenAgreement: PropTypes.string,
  tradeCategory: PropTypes.shape({
    id: PropTypes.number,
    category: PropTypes.string
  }),
  categoryDSelected: PropTypes.bool,
  loggedInUser: PropTypes.shape({
    isGovernmentUser: PropTypes.bool
  })
}

export default CreditTransferSigningHistory
