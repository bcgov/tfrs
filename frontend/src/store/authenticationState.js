import { loadUser } from 'redux-oidc';
import { put, takeLatest, all } from 'redux-saga/effects';
import userManager from './oidc-usermanager';
import { getLoggedInUser } from '../actions/userActions';
import CONFIG from '../config';

const LOGIN_TRIGGERING_ACTIONS = [
  'redux-oidc/USER_EXPIRED'
];

function triggerLoginFlow (store) {
  const { routing } = store.getState();

  if (routing.location &&
    routing.location.pathname !== '/authCallback') {
    return CONFIG.KEYCLOAK.CUSTOM_LOGIN ? userManager.signinSilent() : userManager.signinRedirect();
  }

  return false;
}

function * getBackendUser () {
  yield put(getLoggedInUser());
}

export default function * authenticationStateSaga (store) {
  userManager.clearStaleState();

  const { routing } = store.getState();

  if (!routing.location || routing.location.pathname !== '/authCallback') {
    loadUser(store, userManager);
  }
  yield all([
    takeLatest('redux-oidc/USER_FOUND', getBackendUser),
    takeLatest(
      action => (LOGIN_TRIGGERING_ACTIONS.includes(action.type)),
      triggerLoginFlow,
      store
    )
  ]);
}
