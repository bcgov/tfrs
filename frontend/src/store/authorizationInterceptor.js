import axios from 'axios';
import userManager from './oidc-usermanager';
import CONFIG from '../config';

function configureAxios () {
  if (CONFIG.KEYCLOAK) {
    console.log('adding token to request');

    axios.interceptors.request.use((config) => {
      const loadconfig = async () => {
        const user = await userManager.getUser();
        return {
          ...config,
          headers: {
            Authorization: `Bearer ${user.id_token}`
          }
        };
      };

      return loadconfig();
    });
  }
}

export default configureAxios;
