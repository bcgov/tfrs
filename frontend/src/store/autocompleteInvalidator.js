import { call, put, takeLatest } from 'redux-saga/effects';
import AutocompleteActionTypes from "../constants/actionTypes/Autocomplete";
import FuelCodeActionTypes from '../constants/actionTypes/FuelCodes';

const INVALIDATING_ACTIONS = [
  FuelCodeActionTypes.ADD_FUEL_CODE,
  FuelCodeActionTypes.RECEIVE_FUEL_CODE,
  FuelCodeActionTypes.RECEIVE_FUEL_CODES,
  FuelCodeActionTypes.UPDATE_FUEL_CODE,
  FuelCodeActionTypes.DELETE_FUEL_CODE
];

function * invalidateCache () {
  yield put({ type: AutocompleteActionTypes.INVALIDATE_AUTOCOMPLETE_CACHE });
}

export default function * sessionTimeoutSaga () {
  yield takeLatest(
    action => (INVALIDATING_ACTIONS.includes(action.type)),
    invalidateCache
  );
}
