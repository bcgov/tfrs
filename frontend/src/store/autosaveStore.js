import {takeEvery, all, call} from 'redux-saga/effects';
import {delay} from "redux-saga";
import axios from "axios";
import * as Routes from "../constants/routes";

function* loadAutosaveData(action) {
  const {key, resolve, reject} = action.payload;

  axios
    .get(`${Routes.BASE_URL}${Routes.AUTOSAVE_API}?key=${key}`)
    .then((response) => {
      try {
        const parsed = JSON.parse(response.data.data);
        resolve(parsed);
      } catch (e) {
        reject(e);
      }
    }).catch((error) => {
    reject(error);
  });
}

function* saveAutosaveData(action) {
  const {key, state, resolve, reject} = action.payload;

  const payload = {
    key,
    data: JSON.stringify(state)
  };

  axios
    .post(`${Routes.BASE_URL}${Routes.AUTOSAVE_API}`, payload)
    .then((response) => {
      resolve();
    }).catch((error) => {
    reject(error);
  });
}

function* clearAutosaveData(action) {
  const {resolve, reject} = action.payload;

  axios
    .post(`${Routes.BASE_URL}${Routes.AUTOSAVE_API}/clear`)
    .then((response) => {
      try {
        resolve();
      } catch (e) {
        reject(e);
      }
    }).catch((error) => {
    reject(error);
  });
}

export default function* autosaveSaga() {

  yield all([
    takeEvery('LOAD_AUTOSAVE_DATA', loadAutosaveData),
    takeEvery('SAVE_AUTOSAVE_DATA', saveAutosaveData),
    takeEvery('CLEAR_AUTOSAVE_DATA', clearAutosaveData)
  ]);
}
