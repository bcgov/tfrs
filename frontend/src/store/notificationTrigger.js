import {call, put, takeLatest} from 'redux-saga/effects';
import {delay} from 'redux-saga';
import {getNotifications, getNotificationsCount} from "../actions/notificationActions";
import UserActionTypes from "../constants/actionTypes/Users";
import NotificationActionTypes from "../constants/actionTypes/Notifications";


const TRIGGERING_ACTIONS = [
  'SERVER_INITIATED_NOTIFICATION_RELOAD',
  UserActionTypes.RECEIVE_LOGGED_IN_USER,
  NotificationActionTypes.SUCCESS_NOTIFICATIONS
];

function* fetchNotifications(store) {
  yield call(delay, 1000); //debounce

  if (store.getState().rootReducer.userRequest.isAuthenticated) {
    yield put(getNotificationsCount());
  }

  if (store.getState().rootReducer.notifications.onNotificationsPage) {
    yield put(getNotifications());
  }

}

export default function* notificationsSaga(store) {
  yield takeLatest(
    action => (TRIGGERING_ACTIONS.includes(action.type)),
    fetchNotifications,
    store
  );
}
