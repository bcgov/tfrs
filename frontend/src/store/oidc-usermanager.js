import { createUserManager } from 'redux-oidc';
import CONFIG from '../config';

const settings = {
  authority: CONFIG.KEYCLOAK.AUTHORITY,
  client_id: CONFIG.KEYCLOAK.CLIENT_ID,
  redirect_uri: CONFIG.KEYCLOAK.CALLBACK_URL,
  automaticSilentRenew: true,
  silent_redirect_uri: `${window.location.protocol}//${window.location.hostname}${window.location.port ? `:${window.location.port}` : ''}/token_renew.html`,
  response_type: 'token id_token'
};

const userManager = createUserManager(settings);

export default userManager;
