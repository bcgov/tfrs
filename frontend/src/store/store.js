import { createStore, compose, applyMiddleware, combineReducers } from 'redux';
import { reducer as toastrReducer } from 'react-redux-toastr';
import { routerReducer, routerMiddleware } from 'react-router-redux';
import { createLogger } from 'redux-logger';
import io from 'socket.io-client';
import thunk from 'redux-thunk';
import rootReducer from '../reducers/reducer';
import { getNotifications } from '../actions/notificationActions';
import { SOCKETIO_URL } from '../constants/routes';

import createOidcMiddleware, { loadUser, processSilentRenew, reducer as OIDCReducer } from 'redux-oidc';
import { persistTargetPathReducer } from '../reducers/persistTargetPathReducer';

import userManager from './oidc-usermanager';
import CONFIG from '../config';
import { getLoggedInUser } from '../actions/userActions';

import persistState from 'redux-localstorage';
import createSocketIoMiddleware from 'redux-socket.io';

const middleware = routerMiddleware(history);
const oidcMiddleware = createOidcMiddleware(userManager);

const socket = io(SOCKETIO_URL);
const socketIoMiddleware = createSocketIoMiddleware(socket, 'socketio/');

const enhancer = compose(persistState(['targetPath'], {key: 'tfrs-state'}));

const combinedReducers = (state = {}, action) => {
  const currentRoute = state.routing || {};
  return {
    toastr: toastrReducer(state.toastr, action),
    oidc: OIDCReducer(state.oidc, action),
    routing: routerReducer(state.routing, action),
    targetPath: persistTargetPathReducer(state.targetPath, {...action, currentRoute}),
    rootReducer: rootReducer(state.rootReducer, action),
  };
};

let allMiddleware = [
  thunk,
  socketIoMiddleware,
  oidcMiddleware,
  middleware
];

if (process.env.NODE_ENV !== 'production') {
  allMiddleware.push(createLogger());
}

const store = createStore(
  combinedReducers,
  applyMiddleware(...allMiddleware),
  enhancer
);

let subscriptionProcessing = false;

store.subscribe(() => {
  const state = store.getState();
  if (!subscriptionProcessing) {
    subscriptionProcessing = true;

    if (state.rootReducer.notifications.serverInitiatedReloadRequested === true) {
      store.dispatch(getNotifications());
    }

    if (CONFIG.KEYCLOAK.ENABLED) {
      if (state.oidc.user &&
        !state.rootReducer.userRequest.isFetching &&
        !state.rootReducer.userRequest.isAuthenticated) {
        store.dispatch(getLoggedInUser());
      }
    }

    subscriptionProcessing = false;
  }
});

if (CONFIG.KEYCLOAK.ENABLED) {
  loadUser(store, userManager);
  processSilentRenew();
}

export default store;
