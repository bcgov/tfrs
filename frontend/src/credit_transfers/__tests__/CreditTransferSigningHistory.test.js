import React from 'react'
import { Provider } from 'react-redux'
import renderer from 'react-test-renderer'
import { BrowserRouter } from 'react-router-dom'
import CreditTransferSigningHistory from '../components/CreditTransferSigningHistory'
import store from '../../store/store'
import moment from 'moment'

describe('CreditTransferSigningHistory', () => {

  let agreementDate, submissionDate;

  beforeEach(() => {
    agreementDate = new Date();
    submissionDate = new Date();
  });

  describe('Component Rendering', () => {

    const historyMock = [{
      user: {
        firstName: 'John',
        lastName: 'Doe',
        organization: {
          name: 'Organization Name'
        }
      },
      status: {
        id: 1
      },
      createTimestamp: new Date().toISOString()
    }]

    test('renders the component correctly', () => {
      const props = {
        history: historyMock,
        dateOfWrittenAgreement: moment().format('YYYY-MM-DD'),
        categoryDSelected: false,
        tradeCategory: {
          id: 1,
          category: 'A'
        }
      }

      const component = renderer.create(
        <BrowserRouter>
          <Provider store={store}>
            <CreditTransferSigningHistory {...props} />
          </Provider>
        </BrowserRouter>
      )

      const tree = component.toJSON()
      expect(tree).toMatchSnapshot()
    })

    test('should render the component with dateOfWrittenAgreement as null', () => {
      const props = {
        history: historyMock,
        dateOfWrittenAgreement: null,
        categoryDSelected: false,
        tradeCategory: null
      }

      const component = renderer.create(
        <BrowserRouter>
          <Provider store={store}>
            <CreditTransferSigningHistory {...props} />
          </Provider>
        </BrowserRouter>
      )

      const tree = component.toJSON()
      expect(tree).toMatchSnapshot()
    })
  });

  describe('Category D selection logic', () => {
    test('returns category D if categoryDSelected is true', () => {
      const result = CreditTransferSigningHistory.calculateTransferCategoryAndNextChange(agreementDate, submissionDate, true, null)
      expect(result).toEqual({ category: 'D', nextChangeInMonths: null })
    })
  });

  describe('When a specific category is selected', () => {
    test('returns 6 months for next change if selected category is A', () => {
      const result = CreditTransferSigningHistory.calculateTransferCategoryAndNextChange(agreementDate, submissionDate, false, 'A');
      expect(result).toEqual({ category: 'A', nextChangeInMonths: 6 });
    });

    test('returns 12 months for next change if selected category is B', () => {
      const result = CreditTransferSigningHistory.calculateTransferCategoryAndNextChange(agreementDate, submissionDate, false, 'B');
      expect(result).toEqual({ category: 'B', nextChangeInMonths: 12 });
    });

    test('returns null for next change if selected category is C', () => {
      const result = CreditTransferSigningHistory.calculateTransferCategoryAndNextChange(agreementDate, submissionDate, false, 'C');
      expect(result).toEqual({ category: 'C', nextChangeInMonths: null });
    });
  });

  describe('Calculating category based on time since agreement', () => {
    test('returns category A if differenceInMonths <= 6', () => {
      agreementDate.setMonth(agreementDate.getMonth() - 5) // Five months ago
      const result = CreditTransferSigningHistory.calculateTransferCategoryAndNextChange(agreementDate, submissionDate, false, null)
      expect(result).toEqual({ category: 'A', nextChangeInMonths: 6 })
    })

    test('returns category B if differenceInMonths > 6 and differenceInMonths <= 12', () => {
      agreementDate.setMonth(agreementDate.getMonth() - 7) // Seven months ago
      const result = CreditTransferSigningHistory.calculateTransferCategoryAndNextChange(agreementDate, submissionDate, false, null)
      expect(result).toEqual({ category: 'B', nextChangeInMonths: 12 })
    })

    test('returns category C if differenceInMonths > 12', () => {
      agreementDate.setFullYear(agreementDate.getFullYear() - 2) // Two years ago
      const result = CreditTransferSigningHistory.calculateTransferCategoryAndNextChange(agreementDate, submissionDate, false, null)
      expect(result).toEqual({ category: 'C', nextChangeInMonths: null })
    })
  });
})
