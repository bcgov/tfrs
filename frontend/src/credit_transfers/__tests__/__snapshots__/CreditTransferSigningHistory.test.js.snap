// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreditTransferSigningHistory renders the component correctly 1`] = `
<div
  className="credit-transfer-signing-history"
>
  <h3
    className="signing-authority-header"
  >
    Transaction History
  </h3>
  <p>
    <li>
      <span>
        Date of written agreement reached between the two organizations: 
      </span>
      <strong>
        January 18, 2024
      </strong>
      <span>
         (proposal falls under 
        <strong>
          Category 
          A
        </strong>
        
      </span>
      <span>
         if approved by: 
        <strong>
          July 18, 2024
        </strong>
        )
      </span>
    </li>
  </p>
  <p>
    <li>
      <strong>
         
      </strong>
       
      <span>
         on 
      </span>
      January 18, 2024
      <span>
         by 
      </span>
      <strong>
         
        John
         
        Doe
      </strong>
       of
      <strong>
         
        Organization Name
         
      </strong>
    </li>
  </p>
</div>
`;

exports[`CreditTransferSigningHistory should render the component with dateOfWrittenAgreement as null 1`] = `
<div
  className="credit-transfer-signing-history"
>
  <h3
    className="signing-authority-header"
  >
    Transaction History
  </h3>
  <p>
    <li>
      <strong>
         
      </strong>
       
      <span>
         on 
      </span>
      January 18, 2024
      <span>
         by 
      </span>
      <strong>
         
        John
         
        Doe
      </strong>
       of
      <strong>
         
        Organization Name
         
      </strong>
    </li>
  </p>
</div>
`;
